docs/docs/use_cases/rag/code_understanding.mdx

```diff
@@ -0,0 +1,337 @@
+});
+```
+
+### Chat
+
+We'll setup our model and memory system just as we'd do for any other chatbot application.
+
+```typescript
+import { ChatOpenAI } from "langchain/chat_models/openai";
+```
+
+```typescript
+const model = new ChatOpenAI({ modelName: "gpt-4" });
+```
+
+We're going to use `BufferMemory` as our memory chain. All this will do is take in inputs/outputs from the LLM and store them in memory.
+
+```typescript
+import { BufferMemory } from "langchain/memory";
+```
+
+```typescript
+const memory = new BufferMemory({
+  returnMessages: true, // Return stored messages as instances of `BaseMessage`
+  memoryKey: "chat_history", // This must match up with our prompt template input variable.
+});
+```
+
+Now we can construct our main sequence of chains. We're going to be building `ConversationalRetrievalChain` using Expression Language.
+
+```typescript
+import {
+  ChatPromptTemplate,
+  MessagesPlaceholder,
+  AIMessagePromptTemplate,
+  HumanMessagePromptTemplate,
+} from "langchain/prompts";
+import { RunnableBranch, RunnableSequence } from "langchain/schema/runnable";
+import { formatDocumentsAsString } from "langchain/util/document";
+import { BaseMessage } from "langchain/schema";
+import { StringOutputParser } from "langchain/schema/output_parser";
+```
+
+## Construct the chain
+
+The meat of this code understanding example will be inside a single `RunnableSequence` chain.
+Here, we'll have a single input parameter for the question, and preform retrieval for context and chat history (if available).
+Then we'll preform the first LLM call to rephrase the users question.
+Finally, using the rephrased question, context and chat history we'll query the LLM to generate the final answer which we'll return to the user.
+
+### Prompts
+
+Step one is to define our prompts. We need two, the first we'll use to rephrase the users question, the second we'll use to combine the documents and question.
+
+Question generator prompt:
+
+```typescript
+const questionGeneratorTemplate = ChatPromptTemplate.fromMessages([
+  AIMessagePromptTemplate.fromTemplate(
+    "Given the following conversation about a codebase and a follow up question, rephrase the follow up question to be a standalone question."
+  ),
+  new MessagesPlaceholder("chat_history"),
+  AIMessagePromptTemplate.fromTemplate(`Follow Up Input: {question}
+Standalone question:`),
+]);
+```
+
+Combine documents prompt:
+
+```typescript
+const combineDocumentsPrompt = ChatPromptTemplate.fromMessages([
+  AIMessagePromptTemplate.fromTemplate(
+    "Use the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.\n\n{context}\n\n"
+  ),
+  new MessagesPlaceholder("chat_history"),
+  HumanMessagePromptTemplate.fromTemplate("Question: {question}"),
+]);
+```
+
+Next, we'll construct our chain.
+
+```typescript
+const conversationalQaChain = RunnableSequence.from([
+  {
```

Comment:
I would say breaking these up into smaller logical sequences might help with readability?

Reply:
Good idea, split into two, one for combining docs and one for the conversational part